crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+
+        let latest = checkpoint
+            .as_ref()
+            .map(|checkpoint| checkpoint.sequence_number());
+
+        match self.heights.entry(peer_id) {
+            Entry::Occupied(mut entry) => {
+                if latest > *entry.get() {
+                    entry.insert(latest);
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(latest);
+            }
+        }
+
+        if let Some(checkpoint) = checkpoint {
+            self.insert_checkpoint(checkpoint);
+        }
+    }
+
+    pub fn cleanup_old_checkpoints(&mut self, sequence_number: CheckpointSequenceNumber) {
+        self.unprocessed_checkpoints
+            .retain(|_digest, checkpoint| checkpoint.sequence_number() > sequence_number);
+        self.sequence_number_to_digest
+            .retain(|&s, _digests| s > sequence_number);
+    }
+
+    pub fn insert_checkpoint(&mut self, checkpoint: Checkpoint) {
+        let digest = checkpoint.digest();
+        let sequence_number = checkpoint.sequence_number();
+        self.unprocessed_checkpoints.insert(digest, checkpoint);
+        self.sequence_number_to_digest
+            .entry(sequence_number)
+            .or_default()
+            .insert(digest);
+    }
+}
+
+#[derive(Clone, Debug)]
+enum StateSyncMessage {
+    StartSyncJob,
+    // Validators will send this to the StateSyncEventLoop
+    // in order to kick off gossipping it to other peers
+    VerifiedCheckpoint(Box<VerifiedCheckpoint>),
+}
+
+struct StateSyncEventLoop<S> {
+    mailbox: mpsc::Receiver<StateSyncMessage>,
+    /// Weak reference to our own mailbox
+    weak_sender: mpsc::WeakSender<StateSyncMessage>,
+
+    tasks: JoinSet<()>,
+    sync_tasks: JoinSet<Option<VerifiedCheckpoint>>,
+
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint_event_sender: broadcast::Sender<VerifiedCheckpoint>,
+    network: anemo::Network,
+}
+
+impl<S> StateSyncEventLoop<S>
+where
+    S: CheckpointStore + Clone + Send + Sync + 'static,
+{
+    // Note: A great deal of care is taken to ensure that all event handlers are non-asynchronous
+    // and that the only "await" points are from the select macro picking which event to handle.
+    // This ensures that the event loop is able to process events at a high speed and reduce the
+    // chance for building up a backlog of events to process.
+    pub async fn start(mut self) {
+        info!("State-Synchronizer started");
+
+        let mut interval = tokio::time::interval(Duration::from_secs(60 * 10));
+        let mut peer_events = {
+            let (subscriber, peers) = self.network.subscribe();
+            for peer_id in peers {
+                self.get_latest_from_peer(peer_id);
+            }
+            subscriber
+        };
+
+        loop {
+            tokio::select! {
+                now = interval.tick() => {
+                    self.handle_tick(now.into_std());
+                }
```

Comment:
looks like some of these select handler blocks end with commas, some not? i guess it must not matter but looks weird to me?